rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/action/TbMsgCountNodeTest.java

```diff
@@ -0,0 +1,129 @@
+import org.thingsboard.server.common.data.msg.TbNodeConnectionType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@Slf4j
+@ExtendWith(MockitoExtension.class)
+public class TbMsgCountNodeTest {
+
+    private final TenantId TENANT_ID = TenantId.fromUUID(UUID.fromString("773ccb0e-e822-44b7-97aa-10de83ad081d"));
+    private final ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private final int msgCountInterval = 1;
+
+
+    private TbMsgCountNode node;
+    private TbMsgCountNodeConfiguration config;
+    private CountDownLatch awaitTellSelfLatch;
+
+    @Mock
+    private TbContext ctxMock;
+
+    @BeforeEach
+    public void setUp() {
+        node = new TbMsgCountNode();
+        config = new TbMsgCountNodeConfiguration();
+    }
+
+    @Test
+    public void givenIncomingMsgs_whenOnMsg_thenSendsMsgWithMsgCount() throws TbNodeException, InterruptedException {
+        config.setInterval(msgCountInterval);
+        config.setTelemetryPrefix("count");
+        var configuration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+
+        int msgCount = 100;
+        awaitTellSelfLatch = new CountDownLatch(1);
+        AtomicInteger currentMsgNumber = new AtomicInteger(0);
+        AtomicBoolean isMsgWithCounterSent = new AtomicBoolean(false);
+
+        invokeTellSelf(isMsgWithCounterSent, config.getInterval());
+        when(ctxMock.getTenantId()).thenReturn(TENANT_ID);
+        when(ctxMock.getServiceId()).thenReturn("tb-rule-engine");
+        TbMsg tickMsg = TbMsg.newMsg(TbMsgType.MSG_COUNT_SELF_MSG, null, TbMsgMetaData.EMPTY, TbMsg.EMPTY_STRING);
+        when(ctxMock.newMsg(any(), any(TbMsgType.class), any(), any(), any(), any())).thenReturn(tickMsg);
+
+        node.init(ctxMock, configuration);
+
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, null, TbMsgMetaData.EMPTY, TbMsg.EMPTY_STRING);
+        for (int i = 0; i < msgCount; i++) {
+            if (isMsgWithCounterSent.get()) {
+                break;
+            }
+            node.onMsg(ctxMock, msg);
+            currentMsgNumber.getAndIncrement();
+        }
+
+        awaitTellSelfLatch.await();
+
+        verify(ctxMock, times(currentMsgNumber.get())).ack(any(TbMsg.class));
```

Comment:
Could you please capture all messages acknowledged and verify that all values from capture are the same as messages processed in the loop added above?